### DISCUSSION TITLE
[20 Marzo 2023] Cache

### DISCUSSION THREAD

**Discussion by Elia-Belli** [sapienzastudentsnetwork/architettura-degli-elaboratori#58] [Cache (M-Z)]
![Esame](https://github.com/Elia-Belli/ae2223/assets/107944476/f260f46b-53b8-49c9-8a5d-b98242384204)
### Soluzione del Prof
Le soluzioni fornite dal professore potrebbero essere incomplete o poco chiare, quindi sono benvenute soluzioni che le completano, oppure soluzioni alternative e/o con spiegazioni più approfondite: [Soluzione.pdf](https://github.com/sapienzastudentsnetwork/archelab2223/files/11666099/AE.2021-22.-.Appello.S2.-.2023-03-20.-.Esercizio.cache.-.Scheda.di.calcolo.pdf)

**Comment by CasuFrost** [comment]
![image](https://github.com/sapienzastudentsnetwork/archelab2223/assets/97389762/88daffe4-2071-43e7-acb7-50bb24411cfd)

**Reply by matypist** [reply]
Nota per chi leggerà questa proposta di soluzione in futuro:
- La dimensione del tag della cache L1 dovrebbe essere di $25$ bit (dato da $32 - m - n - 2 = 32 - 4 - 1 - 2 = 25$)
  - $m$ inteso come dimensione dell'offset di word, $4$ poiché la cache L1 è di $16$ word;
  - $n$ inteso come dimensione dell'indice di linea, $1$ poiché la cache L1 è di $2$ set;
  - $2$ per via della dimensione fissa dell'offset di byte.
 
Per spiegazioni più dettagliate e corrette della soluzione ad ogni punto si veda anche la successiva proposta di soluzione di @FeddyLix17

**Comment by FeddyLix17** [comment]
## 1) Accessi in memoria

dopo essersi ricavati le formule usando un approccio grafico

![image](https://github.com/sapienzastudentsnetwork/archelab2223/assets/66442720/1502db8d-617d-47d4-b742-b6d1361eae77)

si riempie la tabella ottenendo

![image](https://github.com/sapienzastudentsnetwork/archelab2223/assets/66442720/e007aa4e-21f4-4332-b7ea-2cb0b68e166c)

## 2) Dimensioni delle cache

considerando che ogni linea della cache è formata

- dal bit di validità

- dal *Least Recently Used* bit (solo L1)

- dal tag

- dal blocco stesso

si ottiene, rispettivamente per le 2 cache
 
- $\Large \text{L1}$
  
   il numero delle linee è dato dal prodotto tra le vie e i set, da cui

$$
    \Large  (1 + 1 + 25 + 16 * 4 * 8) * (8 * 2) \implies (27 + 512) * 16 \implies  539 * 16 \implies 8624\ \text{bit}
$$

- $\Large \text{L2}$

$$
    \Large  (1 + 20 + 128 * 4 * 8) * 8 \implies (21 + 4096) * 8 \implies  4117 * 8 \implies 32936\ \text{bit}
$$

## 3) Tempi della sequenza di accessi

### (a) Tempo totale sequenza di accessi

rielencandoli brevemente dalla foto superiore, si hanno

- 7 accessi in memoria, con tempo $\Large 400 * 7 = 2800\ \text{ns}$

- 3 accessi in L1, con tempo  $\Large 10 * 3 = 30\ \text{ns}$

- 2 accessi in L2, con tempo  $\Large 20 * 2 = 40\ \text{ns}$

$$
    \Large T_{tot} = 2800+ 30 + 40 = 2870\ \text{ns}
$$

### (b) Tempo medio sequenza di accessi

$$
    \Large T_{med} = \frac{\text{T.\ Totale}}{\text{Num.\ accessi}} = \frac{2870}{12} = 239.17\ \text{ns}
$$

### (c) Istruzioni svolte nel tempo medio

$$
    \Large \text{Istruzioni\ eseguite}  = \frac{\text{T.\ medio} *\ \text{Frequenza}}{\text{Clock\ per\ Istruction}} = \frac{239.17 * 24}{16} = 358.75
$$

## 4) Word offset indirizzo 8680 Cache L2

$$
    \Large \text{Word Offset} = (\text{Address} \gg \text{Byte Offset})\ \\%\ 2^{\text{bit Word Offset}}
$$

$$
    \Large  (8680 \gg 2)\ \\%\ 2^{7} \implies 2170\ \\%\ 128 \implies 122
$$

## 5) Pagine Virtuali

ricavandosi la formula usando un approccio grafico

![image](https://github.com/sapienzastudentsnetwork/archelab2223/assets/66442720/503d57b7-32bb-45de-8acd-b065b2330549)

si riempie la tabella ottenendo

![image](https://github.com/sapienzastudentsnetwork/archelab2223/assets/66442720/89214063-ba95-444a-b067-a079a04853e2)
